private/buf/cmd/buf/command/beta/studioagent/studioagent.go

```diff
@@ -0,0 +1,233 @@
+		nil,
+		`The header names that are disallowed by this agent. When the agent receives an enveloped request with these headers set, it will return an error rather than forward the request to the target server. Multiple headers are appended if specified multiple times.`,
+	)
+	flagSet.StringToStringVar(
+		&f.ForwardHeaders,
+		forwardHeadersFlagName,
+		nil,
+		`The headers to be forwarded via the agent to the target server. Must be an equals sign separated key-value pair (like --forward-header=fromHeader1=toHeader1). Multiple header pairs are appended if specified multiple times.`,
+	)
+	flagSet.StringVar(
+		&f.CACert,
+		caCertFlagName,
+		"",
+		"The CA cert to be used in the client and server TLS configuration.",
+	)
+	flagSet.StringVar(
+		&f.ClientCert,
+		clientCertFlagName,
+		"",
+		"The cert to be used in the client TLS configuration.",
+	)
+	flagSet.StringVar(
+		&f.ClientKey,
+		clientKeyFlagName,
+		"",
+		"The key to be used in the client TLS configuration.",
+	)
+	flagSet.StringVar(
+		&f.ServerCert,
+		serverCertFlagName,
+		"",
+		"The cert to be used in the server TLS configuration.",
+	)
+	flagSet.StringVar(
+		&f.ServerKey,
+		serverKeyFlagName,
+		"",
+		"The key to be used in the server TLS configuration.",
+	)
+}
+
+func run(
+	ctx context.Context,
+	container appflag.Container,
+	flags *flags,
+) error {
+	// CA cert pool is optional. If it is nil, TLS uses the host's root CA set.
+	var rootCAConfig *tls.Config
+	var err error
+	if flags.CACert != "" {
+		rootCAConfig, err = certclient.NewClientTLSConfigFromRootCertFiles(flags.CACert)
+		if err != nil {
+			return err
+		}
+	}
+	// client TLS config is optional. If it is nil, it uses the default configuration from http2.Transport.
+	var clientTLSConfig *tls.Config
+	if flags.ClientCert != "" || flags.ClientKey != "" {
+		clientTLSConfig, err = newTLSConfig(rootCAConfig, flags.ClientCert, flags.ClientKey)
+		if err != nil {
+			return fmt.Errorf("cannot create new client TLS config: %w", err)
+		}
+	}
+	// server TLS config is optional. If it is nil, we serve with a h2c handler.
+	var serverTLSConfig *tls.Config
+	if flags.ServerCert != "" || flags.ServerKey != "" {
+		serverTLSConfig, err = newTLSConfig(rootCAConfig, flags.ServerCert, flags.ServerKey)
+		if err != nil {
+			return fmt.Errorf("cannot create new server TLS config: %w", err)
+		}
+	}
+	mux := bufstudioagent.NewHandler(
+		container.Logger(),
+		container.Arg(0), // the origin from command argument
+		clientTLSConfig,
+		stringutil.SliceToMap(flags.DisallowedHeaders),
+		flags.ForwardHeaders,
+	)
+	if serverTLSConfig == nil {
+		mux = h2c.NewHandler(mux, &http2.Server{
+			IdleTimeout: defaultIdleTimeout,
+		})
+	}
+	httpServer := &http.Server{
+		Handler:           mux,
+		ReadHeaderTimeout: defaultReadHeaderTimeout,
+		TLSConfig:         serverTLSConfig,
+	}
+	var httpListenConfig net.ListenConfig
+	httpListener, err := httpListenConfig.Listen(ctx, "tcp", fmt.Sprintf("0.0.0.0:%s", flags.Port))
```

Comment:
I'm thinking we may want an option to let users specify `127.0.0.1` here, rather than hardcode 
`0.0.0.0`. The latter causes the listener to bind to the public interfaces (and makes it reachable outside the machine) whereas 127 will bind on the loopback interface and only be exposed locally. For the case of running the agent in a network the `0.0.0.0` interface should be used, but during local dev `127.0.0.1` is more appropriate.

Let's flag that defaults to `127.0.0.1` that can optionally be set to `0.0.0.0`. For the name I think we can use `bind`, which seems to be used by grpcui, a few k8s tools and a few hashicorp tools.

Lets also add a log statementhere when the http server starts so the user is aware of what we bound to.